### PR TITLE
Implementation proposal for controlplane kind

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature user

```

### CODE REVIEW

File: pkg/cmd/target/target.go [hunk 1]
```diff
@@ -90,14 +87,12 @@ func runCmdTarget(f util.Factory, o *TargetOptions) error {
 	}
 
 	if o.Output == "" {
-		if o.Kind != "" {
+		if o.Kind == TargetKindControlPlane {
+			fmt.Fprintf(o.IOStreams.Out, "Successfully targeted control plane for shoot %q\n", currentTarget.ShootName())
```
Comment (petersutter) [inline pkg/cmd/target/target.go:91]: maybe
```suggestion
			fmt.Fprintf(o.IOStreams.Out, "Successfully targeted control plane of shoot %q\n", currentTarget.ShootName())
```

File: pkg/cmd/target/target.go [hunk 2]
```diff
@@ -249,14 +246,12 @@ func (o *TargetOptions) Complete(f util.Factory, cmd *cobra.Command, args []stri
...
 
-	o.ControlPlane = tf.ControlPlane()
-
 	return nil
 }
 
 // Validate validates the provided options
 func (o *TargetOptions) Validate() error {
-	if o.ControlPlane {
+	if o.Kind == TargetKindControlPlane {
```
Comment (petersutter) [inline pkg/cmd/target/target.go:254]: let's discuss about the change in the validate function (in general)

File: pkg/target/manager.go [hunk 3]
```diff
@@ -65,6 +66,8 @@ type Manager interface {
 	UnsetTargetSeed() (string, error)
 	// UnsetTargetShoot unsets the garden shoot configuration
 	UnsetTargetShoot() (string, error)
+	// UnsetTargetControlPlane unsets the garden shoot configuration
```
Comment (petersutter) [inline pkg/target/manager.go:69]: the go-doc does not fit very vell, or not?

File: pkg/cmd/target/target.go (hunk 2 above)
Comment (holgerkoser) [reply pkg/cmd/target/target.go:254]: I agree. I propose:
```go
switch o.Kind {
case TargetKindControlPlane:
  break
case "":
  return error.New("error")  
default: 
  if o.TargetName == "" {
    return error.New("error")  
  } 
}
```

File: pkg/cmd/target/unset_test.go [hunk 4]
```diff
@@ -211,6 +187,35 @@ var _ = Describe("Command", func() {
...
+
+		fakeGardenClient := fake.NewClientBuilder().WithObjects(project, shoot).Build()
+
+		// setup command
+		targetProvider := internalfake.NewFakeTargetProvider(currentTarget)
+		clientProvider := internalfake.NewFakeClientProvider()
+		clientProvider.WithClient(gardenKubeconfig, fakeGardenClient)
+
+		factory := internalfake.NewFakeFactory(cfg, nil, clientProvider, nil, targetProvider)
+		cmd := cmdtarget.NewCmdUnset(factory, cmdtarget.NewUnsetOptions(streams))
```
Comment (holgerkoser) [inline pkg/cmd/target/unset_test.go:205]: This code is repeated in every single test. It should be moved into a `JustBeforeEach` hook. Maybe not as part of this PR but in general.

File: pkg/target/manager_test.go [hunk 5]
```diff
@@ -347,6 +347,22 @@ var _ = Describe("Manager", func() {
 		assertTargetProvider(targetProvider, target.NewTarget(gardenName, prod1Project.Name, "", ""))
 	})
 
+	It("should be able to target control plane for a shoot", func() {
+		t := target.NewTarget(gardenName, prod1Project.Name, "", prod1GoldenShoot.Name)
+		manager, targetProvider := createTestManager(t, *cfg, clientProvider, kubeconfigCache)
+
+		Expect(manager.TargetControlPlane(ctx)).To(Succeed())
+		assertTargetProvider(targetProvider, target.NewTarget(gardenName, prod1Project.Name, "", prod1GoldenShoot.Name).WithControlPlane(true))
```
Comment (holgerkoser) [inline pkg/target/manager_test.go:355]:
```suggestion
		assertTargetProvider(targetProvider, t.WithControlPlane(true))
```

File: pkg/target/manager_test.go [hunk 6]
```diff
@@ -462,6 +478,24 @@ var _ = Describe("Manager", func() {
 		assertTargetProvider(targetProvider, t)
 	})
 
+	It("should be able to unset selected control plane", func() {
+		t := target.NewTarget(gardenName, prod1Project.Name, "", prod1AmbiguousShoot.Name).WithControlPlane(true)
+		manager, targetProvider := createTestManager(t, *cfg, clientProvider, kubeconfigCache)
+
+		Expect(manager.UnsetTargetControlPlane()).Should(Equal(prod1AmbiguousShoot.Name))
+		assertTargetProvider(targetProvider, target.NewTarget(gardenName, prod1Project.Name, "", prod1AmbiguousShoot.Name))
```
Comment (holgerkoser) [inline pkg/target/manager_test.go:486]:
```suggestion
		assertTargetProvider(targetProvider, t..WithControlPlane(true))
```

File: pkg/target/manager_test.go [hunk 7]
```diff
@@ -462,6 +478,24 @@ var _ = Describe("Manager", func() {
...
+	})
+
+	It("should fail if no control plane targeted", func() {
```
Comment (holgerkoser) [inline pkg/target/manager_test.go:489]: Shouldn't this be idempotent and not return an error?

File: pkg/cmd/target/unset_test.go (hunk 4 above)
Comment (grolu) [reply pkg/cmd/target/unset_test.go:205]: Yes maybe you saw that I already cleaned up this file, also all the test data has been redeclared in every single test. I moved this in the before each function, however skipped the other lines (for now) as I did not want to change the implementation itself. Also, this pattern is repeated in other files e.g. target_test.go. Let's do this in a separate PR.

File: pkg/target/manager_test.go (hunk 7 above)
Comment (grolu) [reply pkg/target/manager_test.go:489]: All other unset operations return an error in this case. E.g. `gctl target unset shoot` also fails if no shoot is set. We would need to change it for these cases, too.

Comment (holgerkoser) [reply pkg/target/manager_test.go:489]: Ok. I would not return error in all these cases but more important is a consistent behavior.

File: pkg/cmd/target/unset_test.go (hunk 4 above)
Comment (holgerkoser) [reply pkg/cmd/target/unset_test.go:205]: 🆗

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Pull request merged**